pkg/teeth/publish_logs.go

```diff
@@ -87,5 +99,59 @@ func SetUpPublisherSourceQuery(ctx context.Context, bqClient BigQueryClient) (*b
 	if err = tmpl.Execute(&b, bqClient.Config()); err != nil {
 		return nil, fmt.Errorf("failed to execute sql template: %w", err)
 	}
-	return bqClient.Query(b.String()), nil
+	return bqClient.Query(ctx, b.String()), nil
+}
+
+// SaveInvocationCommentStatus inserts the statuses into the
+// InvocationCommentStatus table. It uses the batching parameters
+// of the BigQuery client to configure batch inserts.
+//
+// Returns joined errors where each error corresponds to the batch
+// insert in order.
+func SaveInvocationCommentStatus(ctx context.Context, bqClient BigQueryClient, statuses []*InvocationCommentStatusRecord) error {
+	bqInsertBatchSize := int(bqClient.InsertBatchSize())
+	errs := make([]error, 0)
+	for i := 0; i <= len(statuses)/bqInsertBatchSize; i++ {
+		// Wait to prevent being rate-limited by BigQuery insert API.
+		if i > 0 {
+			time.Sleep(bqClient.InsertBatchWait())
+		}
+		start := i * bqInsertBatchSize
+		end := min((i+1)*bqInsertBatchSize, len(statuses))
+		if err := bqClient.Insert(ctx, statuses[start:end]); err != nil {
+			if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+				errs = append(errs, err)
+				break
```

Comment:
On these errors, we want to abort the insertion right away. Why?

Reply:
On context cancellation or timeout, isn't it expected that the calling function should stop and clean up immediately ... therefore `break`? Let me know if I didn't get your point.

Reply:
Ended up removing.